File: app/javascript/embeds/apple_music.js
import MediaSite from "media_site"

class AppleMusic extends MediaSite {
  constructor() {
    super(
      /(https:\/\/(music|podcasts)\.apple\.com\/(..)\/(album|playlist|podcast)\/[^\/]+\/(id)?([0-9]+))(\?.*)?/,

      async (url) => {
        const match = url.match(this.regex)
        if (match) {
          const country = match[1]
          const type = match[2]
          const id = match[3]
          const track_id = match[4]
          return `https://embed.music.apple.com/${country}/${type}/${id}${track_id}`
        }
        return null
      },

      (embedUrl) => {
        const iframe = document.createElement('iframe')
        iframe.src = embedUrl
        iframe.width = "100%"

        const isTrack = /\?i=\d+$/.test(embedUrl)
        iframe.height = isTrack ? "175" : "450"
        iframe.allow = "autoplay *; encrypted-media *; fullscreen *"
        iframe.loading = "lazy"
        iframe.style.borderRadius = "12px"
        iframe.style.border = "0"
        iframe.style.maxWidth = "660px"
        return iframe
      }
    )
  }
}

export default AppleMusic

File: app/models/media_embed/providers/apple_music.rb
module MediaEmbed
  module Providers
    class AppleMusic < Base
      REGEX = %r{\Ahttps://(?:music|podcasts)\.apple\.com/([a-z]{2})/(album|song|playlist|podcast)/[^/]+/(?:id)?([0-9]+)(\?i=[0-9]+)?}i

      def render(url)
        match = url.match(REGEX)
        return unless match

        country, type, id, track_id = match.captures
        embed_url = "https://embed.music.apple.com/#{country}/#{type}/#{id}#{track_id}"
        is_track = track_id.present? || type == "song"

        iframe(embed_url,
          width: "100%",
          height: is_track ? "175" : "450",
          allow: "autoplay *; encrypted-media *; fullscreen *",
          style: style("border-radius: 12px", "border: 0", "max-width: 660px", "margin-inline: auto"))
      end
    end
  end
end
